network/src/legacy/gossip/attestation.rs

```diff
@@ -113,27 +115,43 @@ impl Default for View {
 impl View {
 	fn leaf_view(&self, relay_chain_leaf: &Hash) -> Option<&LeafView> {
 		self.leaf_work.iter()
-			.find_map(|&(ref h, ref leaf)| if h == relay_chain_leaf { Some(leaf) } else { None } )
+			.find_map(|&(ref signing_context, ref leaf)|
+				if signing_context.parent_hash == *relay_chain_leaf { Some(leaf) } else { None }
+			)
+	}
+
+	fn signing_context(&self, relay_chain_leaf: &Hash) -> Option<&SigningContext<Hash>> {
+		self.leaf_work.iter()
+			.find_map(|&(ref signing_context, _)|
+				if signing_context.parent_hash == *relay_chain_leaf { Some(signing_context) } else { None }
+			)
 	}
 
 	fn leaf_view_mut(&mut self, relay_chain_leaf: &Hash) -> Option<&mut LeafView> {
 		self.leaf_work.iter_mut()
-			.find_map(|&mut (ref h, ref mut leaf)| if h == relay_chain_leaf { Some(leaf) } else { None } )
+			.find_map(|&mut (ref signing_context, ref mut leaf)|
+				if signing_context.parent_hash == *relay_chain_leaf { Some(leaf) } else { None }
+			)
 	}
 
 	/// Get our leaves-set. Guaranteed to have length <= MAX_CHAIN_HEADS.
 	pub(super) fn neighbor_info<'a>(&'a self) -> impl Iterator<Item=Hash> + 'a + Clone {
-		self.leaf_work.iter().take(MAX_CHAIN_HEADS).map(|(p, _)| p.clone())
+		self.leaf_work.iter().take(MAX_CHAIN_HEADS).map(|(p, _)| p.parent_hash.clone())
 	}
 
 	/// Note new leaf in our local view and validation data necessary to check signatures
 	/// of statements issued under this leaf.
 	///
 	/// This will be pruned later on a call to `prune_old_leaves`, when this leaf
 	/// is not a leaf anymore.
-	pub(super) fn new_local_leaf(&mut self, relay_chain_leaf: Hash, validation_data: MessageValidationData) {
+	pub(super) fn new_local_leaf(
+		&mut self,
+		signing_context: SigningContext<Hash>,
+		validation_data: MessageValidationData,
```

Comment:
nit: The `signing_context` should go in the `MessageValidationData` because the purpose of that struct is to hold everything necessary to validate messages